### PR TITLE
Readable Prettyprinter for PIR

### LINKED ISSUE #4572: Improve prettyprinters for PIR
### Describe the feature you'd like

There are no readable (non-sexpr) prettyprinters for PIR. Having one would be nice.

### Describe alternatives you've considered

_No response_

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
This fixes #4572.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [X] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [X] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

**Comment by MaximilianAlgehed** [comment]
@effectfully Could you have a look at this?

**Comment by MaximilianAlgehed** [comment]
@michaelpj FYI running `updateMaterialized` locally failed spectacularly on this branch for god knows what reason. This might cause interesting issues in CI - would be good to have an eye on them.

**Comment by effectfully** [comment]
> @effectfully Could you have a look at this?

It's rather hard to keep all the precedence and associativity stuff in one's head when reading pretty-printing code. Could you commit some pretty-printed PIR programs as golden tests?

The code looks good, so unless there's something horribly off in pretty-printed programs (which I doubt), you get my LGTM.

**Comment by MaximilianAlgehed** [comment]
@effectfully I've added some tests so when this passes CI I think it should be ready to merge.

**Comment by MaximilianAlgehed** [comment]
@michaelpj who is responsible for merging this?

File: plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden
Comment (michaelpj) [inline plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden:8]: I think not indenting subsequent lets is a good idea, which seems to happen on line 4 but not here for some reason?

File: plutus-core/plutus-ir/test/prettyprintingReadable/even3Eval.golden
Comment (michaelpj) [inline plutus-core/plutus-ir/test/prettyprintingReadable/even3Eval.golden:8]: ditto

File: plutus-core/plutus-ir/test/prettyprintingReadable/even3Eval
Comment (michaelpj) [inline plutus-core/plutus-ir/test/prettyprintingReadable/even3Eval:1]: this is the same test as even3 :) we just actually evaluate it in the other contexts. no point duplicating it here

File: plutus-core/plutus-ir/test/prettyprintingReadable/listMatchEval
Comment (michaelpj) [inline plutus-core/plutus-ir/test/prettyprintingReadable/listMatchEval:1]: ditto

File: plutus-core/plutus-ir/test/prettyprintingReadable/listMatch.golden
Comment (michaelpj) [inline plutus-core/plutus-ir/test/prettyprintingReadable/listMatch.golden:5]: "readable" :laughing: To be fair, this one is basically impossible to print "nicely".

File: plutus-core/plutus-ir/test/prettyprintingReadable/mutuallyRecursiveTypes.golden
Comment (michaelpj) [inline plutus-core/plutus-ir/test/prettyprintingReadable/mutuallyRecursiveTypes.golden:1]: no need for a test that evaluates to an error!

File: plutus-core/plutus-ir/test/prettyprintingReadable/stupidZero.golden
Comment (michaelpj) [inline plutus-core/plutus-ir/test/prettyprintingReadable/stupidZero.golden:5]: again increasing indentation for some reason?

File: plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden
Comment (michaelpj) [inline plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden:4]: What do you think about this layout?
```
let ....
in
let ...
in
...
```
Pros: `let` always appears at the start of the line, easier to scan
Cons: more lines

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:29]: Haddock. I'm surprised we don't have this already somewhere...

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:11]: I find it suspicious that this module has no comments :)

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:37]: what does this do? haddock

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:40]: haddock!

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:37]: We have a module for prettyprinter extras: `Data.Prettyprinter.Custom`, if these are generically useful then put them there.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:66]: isn't this `viewApp`?

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:71]: I agree this is impossible, but it doesn't hurt to be specific. "Impossible: this case should be covered by the 'appLike' guard"

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:76]: another prime view pattern opportunity

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:56]: if you made `viewApp` return `Maybe` you could do this as a view pattern

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:76]: I'll stop saying this but you get the idea.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:80]: Should we be exporting something like `typeBinderDocM` for term bindings? @effectfully ?

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:101]: See, I'm sure you thought hard about where to put all these `align`s and so on, but all that knowledge is about to be lost because you haven't written it down :)

**Review by michaelpj — CHANGES_REQUESTED** [review]
I'd like @effectfully to take another look now there's some example output. Mostly just nits and I would really like some comments.

File: plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden
Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden:4]: Other than being a bit (a lot) unorthodox I can't say I particularly mind it. Your call.

Comment (michaelpj) [reply plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden:4]: I think we can keep the orthodox version, it just occurred to me!

File: plutus-core/plutus-ir/test/prettyprintingReadable/listMatch.golden
Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/test/prettyprintingReadable/listMatch.golden:5]: We had a version that would print this with unicode foralls and wouldn't print `*`. That's a lot more readable but in this version I wanted to make sure we at least try to remain somewhat uniform with the other prettyprinters.

Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/test/prettyprintingReadable/listMatch.golden:5]: What do you think about the current version of how we pretty-print type binders here? (i.e. not showing `*`). I would like to make this the default for types as well, but then we'll have to do some work elsewhere too.

Comment (effectfully) [reply plutus-core/plutus-ir/test/prettyprintingReadable/listMatch.golden:5]: `PlutusCore.Pretty.Readable` has 

```haskell
data ShowKinds
    = ShowKindsYes
    | ShowKindsNo
```

which determines whether to show kinds or not. We use it here:

```haskell
-- | Pretty-print a binding at the type level.
typeBinderDocM
    :: ( MonadReader env m, HasPrettyConfigReadable env configName
       , PrettyReadableBy configName tyname
       )
    => ((tyname -> Kind a -> Doc ann) -> AnyToDoc (PrettyConfigReadable configName) ann -> Doc ann)
    -> m (Doc ann)
typeBinderDocM k = do
    showKinds <- view $ prettyConfig . pcrShowKinds
    withPrettyAt ToTheRight botFixity $ \prettyBot -> do
        let prettyBind name kind = case showKinds of
                ShowKindsYes -> parens $ prettyBot name <+> "::" <+> prettyBot kind
                ShowKindsNo  -> prettyBot name
        encloseM binderFixity $ k prettyBind prettyBot
```

which omits kind annotations in binders when `ShowKindsNo` is chosen.

Do you think you could use this existing machinery?

Comment (effectfully) [reply plutus-core/plutus-ir/test/prettyprintingReadable/listMatch.golden:5]: `typeBinderDocM` is not exported from `PlutusCore.Core.Instance.Pretty.Readable`, but do feel free to export and use it, if appropriate.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (effectfully) [reply plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:103]: Yup. Left a comment elsewhere.

File: plutus-core/plutus-ir/test/prettyprintingReadable/listMatch.golden
Comment (effectfully) [reply plutus-core/plutus-ir/test/prettyprintingReadable/listMatch.golden:5]: Aha, just saw that you've already done that.

Comment (michaelpj) [reply plutus-core/plutus-ir/test/prettyprintingReadable/listMatch.golden:5]: Yeah, I'm confused by your comment Max, it looks like you're already using the printer that's sensitive to printing kinds, and also in this example the kinds _are_ printed?

**Comment by MaximilianAlgehed** [comment]
My point is that when the kind is Type you most often don't want to print
it even in situations where you do want to print kinds like Type ->  Type.

The function you mention is all-or-nothing. Having this medium option would
be nice as the default.

File: plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden
Comment (michaelpj) [reply plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden:8]: I still care about this comment! I hadn't quite processed it, but I see that it's related to the previous one: this is consistently aligned with the preceding `let`, so if we did
```
let
in
let
in
```
then we could have the same rule consistently and also get nonincreasing indentation.

I think it is _quite_ desirable to get nonincreasing indendation for nested lets, because PIR programs of any reasonable size tend to have a *lot* of them, is there a way we can do that nicely? If we have to do the unorthodox indentation of the lets then I can live with that.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (effectfully) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:32]: `ShowKindsNo` here.

**Comment by michaelpj** [comment]
> My point is that when the kind is Type you most often don't want to print it even in situations where you do want to print kinds like Type ->  Type.

I'd be happy with adding a `ShowNonTypeKinds` alternative to `ShowKinds`.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:32]: I propose a third `ShowComplexKinds` option that we would use here. I think having no kinds makes complex generated programs (which I care a lot about) almost unreadable.

File: plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden
Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden:8]: I buy having non-increasing indentation being a nice thing. The thing to ensure is that non-let terms below a let should still be indented, otherwise programs just look _weird_. I'll implement this option.

Comment (michaelpj) [reply plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden:8]: The unorthodox one does avoid this by having the body-terms *always* be at the same indentation level. But perhaps it's too weird.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:32]: `instance Default ShowKinds` if you like.

**Comment by effectfully** [comment]
> I'd be happy with adding a `ShowNonTypeKinds` alternative to `ShowKinds`.

Me too.

File: plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden
Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden:8]: Also I'm not sure I see the point about line 4 here. On that line we also indent?

Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden:8]: (Which we would under the orthodox scheme and wouldn't under the unorthodox scheme)

Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden:8]: Yep, `in let` and `letrec` have the same length. Ha.

**Review by effectfully — APPROVED** [review]
The output looks great to me.

I did not go into verifying every single line to check if parens always get generated where they're supposed to (let alone to check if the minimal amount parens get generated), but what I verified did make perfect sense to me.

Some nitpicking below, but feel free to ignore all of it.

File: plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden
Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/test/prettyprintingReadable/even3.golden]: Ok, I've changed this now to gather up the lets and lay them out in a somewhat reasonable fashion as a kind of middle-ground between the unorthodox and orthodox methods that I think looks pretty readable.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs]: I've now made this change and I think it also makes the pretty-printed plc programs a bit more readable.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:92]: I think there are a tonne of different options here and none of them are perfect. What we have now works reasonably well. At some point in the future one might want to re-work it to get rid of some corner-cases. But for now I would vote for keeping what's in the PR.

**Comment by MaximilianAlgehed** [comment]
@michaelpj what do you think about the state of this now?

File: plutus-core/plutus-ir/test/prettyprintingReadable/stupidZero.golden
Comment (michaelpj) [inline plutus-core/plutus-ir/test/prettyprintingReadable/stupidZero.golden:5]: I think this `in` would ideally be on the next line, but I think that would require being fairly clever.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:183]: why isn't this using `typeBinderDocM` any more?

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:37]: TODO: put these in a utility module somewhere, they're generically useful.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:46]: Should probably use `TyVarDecl` and `VarDecl` below, but I can tidy this later.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:114]: This is inconsistent: you use `TyVarDecl` above, but not `VarDecl` here. Both would be simpler if the view functions returned them.

**Review by michaelpj — APPROVED** [review]
Mostly just nits that I'm happy to clean up afterwards. The only question I'm actually unsure about is why you're not using `typeBinderDocM` any more.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:183]: Because `typeBinderDocM` gives you too many brackets as it has to deal with not only the declaration but e.g. the surrounding context (like in a type abstraction).

File: plutus-core/plutus-ir/test/prettyprintingReadable/stupidZero.golden
Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/test/prettyprintingReadable/stupidZero.golden:5]: Yeah this is a tricky one because you want one-liners when you don't have a surrounding context of lets. Maybe not worth the extra complexity in the code.

Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/test/prettyprintingReadable/stupidZero.golden:5]: If one wanted to spend unlimited time tweaking this one could also consider doing something like:
```
let x : t = e
in let
     y : t = e
in let
     z : t = e
in foo x y z
```
or
```
let     x : t = e
in let y : t = e
in let z : t = e
in foo x y z
```
or ...

Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/test/prettyprintingReadable/stupidZero.golden:5]: But I think we are at a good point with this where the marginal value of another hour spent on this is lower than a bunch of other stuff that needs doing so I'm pretty happy.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:37]: I'll let you decide where to put them.

File: plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs
Comment (MaximilianAlgehed) [reply plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs:182]: Also, that function allows for more arguments than we need here.

**Pull request merged**
